From 9e93c1e6e5f6aff12c8c982e96b83f43382b2b8e Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Sun, 17 Mar 2019 18:35:07 +0100
Subject: [PATCH] arm64: irq_pipeline: fix user return from el0_irq

w0 might be trashed as a result of calling
trace_hardirqs_on_pipelined, leading to the wrong decision to branch
to the in-band epilogue when leaving the interrupt frame, although the
in-band stage was stalled on entry.
---
 arch/arm64/kernel/entry.S | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/arch/arm64/kernel/entry.S b/arch/arm64/kernel/entry.S
index fe728839943a..171ad3b82a72 100644
--- a/arch/arm64/kernel/entry.S
+++ b/arch/arm64/kernel/entry.S
@@ -930,16 +930,24 @@ el0_irq_naked:
 1:
 #endif
 	irq_handler
+#ifdef CONFIG_IRQ_PIPELINE
+	cbz	w0, work_done_el0
+#endif
 #ifdef CONFIG_TRACE_IRQFLAGS
 	bl	trace_hardirqs_on_pipelined
-#endif
-#ifdef CONFIG_IRQ_PIPELINE
-	cbz	w0, work_done
 #endif
 	disable_daif
 	b	ret_to_user_noirq
 ENDPROC(el0_irq)
 
+#ifdef CONFIG_IRQ_PIPELINE
+work_done_el0:
+#ifdef CONFIG_TRACE_IRQFLAGS
+	bl	trace_hardirqs_on_pipelined
+#endif
+	b	work_done
+#endif
+
 el1_error:
 	kernel_entry 1
 	mrs	x1, esr_el1
-- 
2.16.4

